src/placement.c

```diff
@@ -0,0 +1,513 @@
+{
+	/*
+	 * The number of row or column intervals is one less than corresponding
+	 * number of row or column grid points.
+	 */
+	int nri = bmp->nrows - 1;
+	int nci = bmp->ncols - 1;
+
+	int i_incr = down ? 1 : -1;
+	int j_incr = right ? 1 : -1;
+
+	int overlap = 0;
+	int count = 0;
+
+	/* Walk up or down along rows according to preference */
+	for (int ii = i; ii >= 0 && ii < nri && height > 0; ii += i_incr) {
+		/* Height of this row */
+		int rh = bmp->rows[ii + 1] - bmp->rows[ii];
+
+		/* Height of overlap between this row and test region */
+		int mh = MAX(0, MIN(height, rh));
+
+		/* Remaining height to consider for next row */
+		height -= rh;
+
+		/* Walk left or right along columns according to preference */
+		int ww = width;
+		for (int jj = j; jj >= 0 && jj < nci && ww > 0; jj += j_incr) {
+			/* Width of this column */
+			int cw = bmp->cols[jj + 1] - bmp->cols[jj];
+
+			/* Width of overlap between this column and test region */
+			int mw = MAX(0, MIN(ww, cw));
+
+			/* Add overlap contribution for this interval */
+			overlap += overlap_bitmap_index(bmp, ii, jj) * mh * mw;
+
+			/* Count the number of overlapping intervals */
+			count++;
+
+			/* Remaining width to consider for next column */
+			ww -= cw;
+		}
+
+		/*
+		 * If there is width left to consider after walking columns,
+		 * the region extends out of bounds and placement is invalid.
+		 */
+		if (ww > 0) {
+			overlap = INT_MAX;
+			break;
+		}
+	}
+
+	/*
+	 * If there is height left ot consider after walking rows, the region
+	 * extends out of bounds and placement is invalid.
+	 */
+	if (height > 0) {
+		overlap = INT_MAX;
+	}
+
+	/* Indicate whether overlap is confined to a single region */
+	if (single) {
+		*single = (count == 1);
+	}
+
+	return overlap;
+}
+
+/*
+ * Find, in (*x, *y), the placement of *view on its output that will minimize
+ * overlap with all other views.
+ */
+bool
+placement_find_best(struct view *view, int *x, int *y)
+{
+	assert(view);
+
+	struct server *server = view->server;
+	assert(server);
+
+	struct border margin = ssd_get_margin(view->ssd);
+
+	struct output *output = view->output;
+	if (!output_is_usable(output)) {
+		return false;
+	}
+
+	/* Default placement is just the upper-left corner of the output */
+	struct wlr_box usable = output_usable_area_in_layout_coords(output);
+	*x = usable.x + margin.left;
+	*y = usable.y + margin.top;
+
+	/* Build the placement grid and overlap bitmap */
+	struct overlap_bitmap bmp = placement_build_bitmap(view);
+
+	int height = view->pending.height + margin.top + margin.bottom;
+	int width = view->pending.width + margin.left + margin.right;
+
+	int min_overlap = INT_MAX;
+
+	int nri = bmp.nrows - 1;
+	int nci = bmp.ncols - 1;
+
+	/* Walk the grid to test all possible placements of the view */
+	for (int i = 0; i < nri; ++i) {
+		for (int j = 0; j < nci; ++j) {
+			/* Search southeast, northeast, northwest, southwest */
+			for (int ii = 0; ii < 4; ++ii) {
+				bool rt = (ii & 0x1) == 0;
```

Comment:
It's not obvious to me how this works. A comment or long variable names or something like that would be helpful.